Project 9

Aivaxela opened the pull request:
project 9 submission 1 ready

Aivaxela committed:
get token and begin brief

Aivaxela committed:
create api class, test pulling user data

Aivaxela committed:
more structuring on Api class

Aivaxela committed:
profile name/about updating from api

Aivaxela committed:
profile name/about setting/getting complete

Aivaxela committed:
refactor api class and add const api data

Aivaxela committed:
trial and error w getting intial cards and adding new

Aivaxela committed:
get init cards, add new card, delete card all functional

Aivaxela committed:
refactored Api class

Aivaxela committed:
adjust modal css for new modal

Aivaxela committed:
start popup confirm logic

Aivaxela committed:
card delete functioning

Aivaxela committed:
like buttons half functional

Aivaxela committed:
card likes are functioning

Aivaxela committed:
refactored index settings, avatar update working

Aivaxela committed:
improved UX, updated gh-pages, ready for submission

Aivaxela committed:
adjusted transistions

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,25 @@
+        }
+        return Promise.reject(`Error: ${res.status}`);
+      })
+      .catch((err) => {
+        console.error(err);
+      });
```
It’s better not to use `catch` in `Api` because you need to `catch` possible errors at the very **end** of any request and this end is in `index.js`  after all blocks `then`. And in real projects blocks `catch` usually open error popups and so on.

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,25 @@
+    this._headers = headers;
+    this._body = body;
+  }
+
```
In `Api` there should be all methods for fetching data: `getInitialCards`, `geUserInfo` and so on and only then they should be called in `index.js`

gennady-bars commented on src/components/Card.js:
```diff
@@ -18,28 +28,45 @@ export default class Card {
+
+  _updateLikeStatus = () => {
+    if (this._isLiked) {
+      this._cardLikeIcon.classList.remove("card__like-button_pressed");
```
You need to **toggle likes** only in block `then` after a successful response from the server because if there is an error the user will not know about that and will see the toggled heart color.

gennady-bars commented on src/components/Card.js:
```diff
@@ -18,28 +28,45 @@ export default class Card {
+      });
+      this._isLiked = false;
+    } else {
+      this._cardLikeIcon.classList.add("card__like-button_pressed");
```
You need to **toggle likes** only in block `then` after a successful response from the server because if there is an error the user will not know about that and will see the toggled heart color.

gennady-bars commented on src/components/Card.js:
```diff
@@ -18,28 +28,45 @@ export default class Card {
+        method: "DELETE",
+        cardEl: this._cardElement,
+      });
+      this._isLiked = false;
```
You need to **toggle likes** only in block `then` after a successful response from the server because if there is an error the user will not know about that and will see the toggled heart color.

gennady-bars commented on src/components/Card.js:
```diff
@@ -18,28 +28,45 @@ export default class Card {
+        method: "PUT",
+        cardEl: this._cardElement,
+      });
+      this._isLiked = true;
```
You need to **toggle likes** only in block `then` after a successful response from the server because if there is an error the user will not know about that and will see the toggled heart color.

gennady-bars commented on src/components/PopupConfirm.js:
```diff
@@ -0,0 +1,34 @@
+    this._data = data;
+  }
+
+  _getInputValues() {
+    const inputFieldValues = {};
+    this._inputFields.forEach((field) => {
+      inputFieldValues[field.name] = field.value;
+    });
+
```
`_getInputValues` is not needed in `PopupConfirm` because you don't have inputs

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -24,11 +25,16 @@ export default class PopupWithForm extends Popup {
-    this._handleFormSubmit(this._getInputValues(), event);
+  closeAfterSubmit() {
     this._popupForm.reset();
+    this._submitButton.textContent = this._submitButtonText;
```
You need to return the **default** button text only in blocks `finally` so that the button could get the text back in any case: `success` or `failure`.

gennady-bars replied:
You need to return the **default** button text only in blocks `finally` so that the button could get the text back in any case: `success` or `failure`.

So, if you call `closeAfterSubmit` in `finally`, then you need to delete `closing` and resetting, `toggleButtonState` from it (they should be in `then`)

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -24,11 +25,16 @@ export default class PopupWithForm extends Popup {
+
+  _onPopupSubmit = (evt) => {
+    const event = evt;
+    this._submitButton.textContent = "Saving...";
```
**Could be improved**

It’s better to make method `renderLoading` inside class `PopupWithForm`. You should find the button once in the constructor and change the text:  

```js
  constructor({ popupSelector, handleSubmit }) {
    super(popupSelector);
    // fix the initial button text only once in the constructor
    this._submitBtnText = this._submitBtn.textContent
  }
  // add 2 params: isLoading and loadingText with a default text
  renderLoading(isLoading, loadingText='Saving...') {
    if (isLoading) {
      this._submitBtn.textContent = loadingText;
    } else {
	// here we return back the initial text. So, you don’t need to bother yourself about it
      this._submitBtn.textContent = this._submitBtnText;
    }
  }

```

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -24,11 +25,16 @@ export default class PopupWithForm extends Popup {
     this._formValidator.toggleButtonState();
     this.close();
+  }
+
```
**Could be improved**

You can make ` setInputValues` in class `PopupWithForm`, which can insert data into inputs:

```js

  setInputValues(data) {
    this._inputList.forEach((input) => {
      // here you insert the `value` by the `name` of the input
      input.value = data[input.name];
    });
  }



```

So, this way you’ll not have to search the inputs of the profile in `index.js`

gennady-bars commented on src/components/UserInfo.js:
```diff
@@ -1,15 +1,38 @@
+import Api from "./Api.js";
```
From the 8th sprint it's not allowed to import anything to the classes. Only class `Popup` can be imported for extending. **`Loose coupling`** should be used: it means that you need to pass callbacks to the constructor rather than using outside elements or functions.

gennady-bars commented on src/components/UserInfo.js:
```diff
@@ -1,15 +1,38 @@
 
   getUserInfo() {
-    return { name: this._name.textContent, title: this._title.textContent };
+    const userInfoApiGet = new Api({
```
Please, don't use api calls inside `UserInfo` because it's only for rendering the user  data on the page

gennady-bars commented on src/components/UserInfo.js:
```diff
@@ -1,15 +1,38 @@
   }
 
   getUserInfo() {
-    return { name: this._name.textContent, title: this._title.textContent };
```
`   return { name: this._name.textContent, title: this._title.textContent };` should be here like it was previously


You don't need to waste resources on fetching the data when you open the profile form

gennady-bars commented on src/pages/index.js:
```diff
@@ -2,46 +2,84 @@
+  apiHeaders: apiData.headers,
+  apiUser: apiData.currentUser,
+});
+userInfo.getUserInfo();
```
You need to `catch` possible errors at the very **end** of any server request.

**Please, correct it everywhere**

gennady-bars commented on src/pages/index.js:
```diff
@@ -2,46 +2,84 @@
+});
+userInfo.getUserInfo();
+
+const initialCardsApi = new Api({
```
`new Api` should be called only **once** because all methods will be ready inside `Api`

According to the task you need to pass full `headers` and `baseUrl` into the constructor of `Api`. See here https://disk.yandex.ru/i/J93dbJp7eypXKQ . 


So, you'll have a `baseUrl` and inside `Api` you'll use it adding **endpoints** to it not to do that in `index.js`


All your methods will have clear understandable names

gennady-bars commented on src/pages/index.js:
```diff
@@ -2,46 +2,84 @@
+  cardsListSection.addItem(createCard(item), firstRender);
+}, cardsListSelector);
+
 const profileFormValidator = new FormValidator(validatorConfig, profileModalForm);
```
**Could be improved**

You can **universally** create instances of validators for all forms in the project storing them inside one object `formValidators`.  And then you can take any validator using attribute `name` of the form where you need to disable/enable the submit button or remove errors.

```js
// define an object for storing validators
const formValidators = {}

const enableValidation = (config) => {
  const formList = Array.from(document.querySelectorAll(config.formSelector))
  formList.forEach((formElement) => {
    const validator = new FormValidator(config, formElement)
    // here you get the name of the form (if you don’t have it then you need to add it into each form in `index.html` first)
    const formName = formElement.getAttribute('name')

   // here you store the validator using the `name` of the form
    formValidators[formName] = validator;
    validator.enableValidation();
  });
};

enableValidation(config);

```

And now you can use them for disabling buttons or clearing errors:

```js

formValidators[ profileForm.getAttribute('name') ].resetValidation()

// or you can use a string – the name of the form (you know it from `index.html`)

formValidators['profile-form'].resetValidation()

```

gennady-bars commented on src/pages/index.js:
```diff
@@ -2,46 +2,84 @@
+  "#avatar-edit-modal",
+  (inputValue, evt) => {
+    evt.preventDefault();
+    profileAvatar.src = inputValue.Link;
```
You need to update `DOM` only in blocks `then` after a successful response from the server so that when there is an error it wouldn’t change

gennady-bars replied:
Class `UserInfo` should set the entire user’s data (including the **avatar**)

gennady-bars commented on src/pages/index.js:
```diff
@@ -2,46 +2,84 @@
     evt.preventDefault();
-    const { Link: link, Title: title } = inputFieldValues;
-    createCard({ name: title, link: link });
+    uploadCard(inputFieldValues);
```
Every server request in `index.js` should look something like this:

```js
api.someMethod()
.then(res => {
// here is the response that should be used to update the DOM.
// also in `then` you should close the popup
})
.catch(err => // here you catch possible errors)
.finally(() => // here you bring the default button text back )
``` 

Please, correct it in the whole code

gennady-bars commented on src/pages/index.js:
```diff
@@ -50,38 +88,88 @@ addImagePopup.setEventListeners();
-  const { name, title } = userInfo.getUserInfo();
-  profileModalNameInput.value = name;
-  profileModalDescInput.value = title;
+  profileModalNameInput.value = userInfo.name.textContent;
```
`userInfo.name.textContent;`

You have `getUserInfo` for that

gennady-bars commented on src/pages/index.js:
```diff
@@ -50,38 +88,88 @@ addImagePopup.setEventListeners();
+    headers: apiData.headers,
+  });
+  deleteCardApi.handleFetch();
+  cardData.cardEl.remove();
+  cardData.cardEl = null;
 }
```
You need to **delete** cards only in block `then` after a **successful** response from the server so that if there is an error the card would stay in the `DOM`.

gennady-bars commented on src/pages/index.js:
```diff
@@ -50,38 +88,88 @@ addImagePopup.setEventListeners();
+    url: apiData.currentUserAvatar,
+    method: "PATCH",
+    headers: apiData.headers,
+    body: JSON.stringify({
```
` body: JSON.stringify({` will be inside `Api` to hide everything related to the api calls inside the class. It's the main responsibility of `Api`. You shouldn't see the whole logics inside `index.js`. You'll see only `api.uploadCard(inputValues)`


```js
// start loading effect
  api
    .uploadCard(inputValues)
    .then((res) => {
      cardsListSection.addItem(createCard({ name: title, link: link, _id: res._id }), false);
    })
    .then(() => addImagePopup.closeAfterSubmit())
    .catch(...) // don't forget about `catch`
    .finally(... stop the loading effect here)
```


**Please, correct it everywhere**

gennady-bars commented on src/pages/index.js:
```diff
@@ -2,46 +2,84 @@
+    evt.preventDefault();
+    profileAvatar.src = inputValue.Link;
+    // evt.submitter.textContent = "Saving...";
+    updateAvatar(profileAvatar.src);
```
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```

gennady-bars reviewed (commented):
Dear Riley , (**Please, expand the general comment ↓**)

Good job but there are some remarks:



* It’s better not to use `catch` in `Api` because you need to `catch` possible errors at the very **end** of any request and this end is in `index.js`  after all blocks `then`. And in real projects blocks `catch` usually open error popups and so on.
* In `Api` there should be all methods for fetching data: `getInitialCards`, `geUserInfo` and so on and only then they should be called in `index.js`
* You need to **toggle likes** only in block `then` after a successful response from the server because if there is an error the user will not know about that and will see the toggled heart color.
* You need to return the **default** button text only in blocks `finally` so that the button could get the text back in any case: `success` or `failure`.
* From the 8th sprint it's not allowed to import anything to the classes. Only class `Popup` can be imported for extending. **`Loose coupling`** should be used: it means that you need to pass callbacks to the constructor rather than using outside elements or functions.
* You need to `catch` possible errors at the very **end** of any server request.
* You need to update `DOM` only in blocks `then` after a successful response from the server so that when there is an error it wouldn’t change
* Class `UserInfo` should set the entire user’s data (including the **avatar**)
* You need to **delete** cards only in block `then` after a **successful** response from the server so that if there is an error the card would stay in the `DOM`.


Please, correct everything and your project will be accepted.

Good luck with the refactoring

Aivaxela committed:
refactoring api calls index.js

Aivaxela committed:
refactoring of api calls and api catch methods completed

gennady-bars reviewed (commented):
Dear Riley , (**Please, expand the general comment ↓**)

Unfortunately I have to reject the project without reviewing:


* You haven't corrected some of my remarks from here https://github.com/Aivaxela/se_project_aroundtheus/pull/6#pullrequestreview-2051150933

Aivaxela committed:
remove import from popup.js, remove comments

Aivaxela commented:
Gennady,
Thanks for reviewing my project!
I am having a hard time finding which remarks I did not address. If you need to reject again, can you please clarify which remarks I am not correcting?

- It’s better not to use catch in Api because you need to catch possible errors at the very end of any request and this end is in index.js after all blocks then. And in real projects blocks catch usually open error popups and so on.
**- I removed all the catches from the Api and added them to index.js after the then() blocks.**
- In Api there should be all methods for fetching data: getInitialCards, geUserInfo and so on and only then they should be called in index.js
**- Api class has all of the methods in it now.**
- You need to toggle likes only in block then after a successful response from the server because if there is an error the user will not know about that and will see the toggled heart color.
**- I tested that the likes do not change color if a server error pops up.**
- You need to return the default button text only in blocks finally so that the button could get the text back in any case: success or failure.
**- I tested that the default button text returns on both server error and success.**
- From the 8th sprint it's not allowed to import anything to the classes. Only class Popup can be imported for extending. Loose coupling should be used: it means that you need to pass callbacks to the constructor rather than using outside elements or functions.
**-I removed the import statement from the Popup class and UserInfo class.**
- You need to catch possible errors at the very end of any server request.
- You need to update DOM only in blocks then after a successful response from the server so that when there is an error it wouldn’t change
- Class UserInfo should set the entire user’s data (including the avatar)
**-Setting the avatar now takes palce in the UserInfo class**
- You need to delete cards only in block then after a successful response from the server so that if there is an error the card would stay in the DOM.
**-I tested that the card will not get deleted if an error occurs**

Thank you!

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,99 @@
+    return fetch(`${this._baseUrl}cards`, {
+      headers: this._headers,
+    }).then((res) => {
+      if (res.ok) {
+        return res.json();
+      }
+      return Promise.reject(`Error: ${res.status}`);
```
```js
.then(res => {
                if (res.ok) {
                    return res.json();
                }
                return Promise.reject(`Error ${res.status}`);
            })

```

The code of `ok` checking is repeated in **every** request of the project. To avoid it you need to make a special method `_checkResponse` that will check any response from the server if it’s ok or not.


```js
_checkResponse(res) {
        // here is the code of the checking
    }

```

And now you can replace redundant repeating with the only line:


```js
 .then(this._checkResponse)

```

Please, pay attention that you need to pass only the **reference** for the method rather than the call. 

**Could be improved**

You can make a special method (function ) for fetching and checking responses **not to duplicate** it in every request:

```js

  _request(url, options) {
    return fetch(url, options).then(this._checkResponse)
  }

```

And now you can safely replace `fetch` with `this._request` and remove repeated checking. And you don’t need to change anything else.

gennady-bars commented on src/components/PopupConfirm.js:
```diff
@@ -0,0 +1,25 @@
+  _submitPressed = (evt) => {
+    const event = evt;
+    this._onConfirm(this._data, event);
+    this.close();
```
You need to close popups only in blocks `then` after a **successful response** from the server so that when there is an error it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well

gennady-bars commented on src/pages/index.js:
```diff
@@ -2,86 +2,166 @@
+  .catch((err) => alert(err));
 
 const imagePopup = new PopupWithImage("#image-modal");
 imagePopup.setEventListeners();
 
-const cardsList = new Section(
-  {
-    data: initialCards,
-    renderer: (item) => {
-      createCard(item);
-    },
-  },
-  cardsListSection
-);
-cardsList.renderItems();
+const confirmPopup = new PopupConfirm("#confirm-modal", (cardData, evt) => {
+  evt.preventDefault();
+  api
+    .deleteCard(cardData.cardId)
+    .then(() => {
+      cardData.cardEl.remove();
+      cardData.cardEl = null;
+    })
+    .catch((err) => alert(err));
+});
+confirmPopup.setEventListeners();
 
 //add event listeners
-profileEditButton.addEventListener("click", openProfileForm);
-profileAddButton.addEventListener("click", openAddForm);
-
-//event listener callbacks
-function openProfileForm() {
+profileEditButton.addEventListener("click", () => {
   profilePopup.open();
-  const { name, title } = userInfo.getUserInfo();
-  profileModalNameInput.value = name;
-  profileModalDescInput.value = title;
+  profileModalNameInput.value = userInfo.getUserInfo().name;
+  profileModalDescInput.value = userInfo.getUserInfo().about;
```
no need to call `userInfo.getUserInfo()` 2 times. Last time you had this code here

```js

  const { name, about} = userInfo.getUserInfo();
  profileModalNameInput.value = name;
  profileModalDescInput.value = about;
```

gennady-bars commented on src/pages/index.js:
```diff
@@ -2,86 +2,166 @@
+      api
+        .updateCardLike(card.id, method)
+        .then((res) => {
+          if (res.isLiked) {
+            card.cardLikeIcon.classList.add("card__like-button_pressed");
+            card.isLiked = true;
+          } else {
+            card.cardLikeIcon.classList.remove("card__like-button_pressed");
+            card.isLiked = false;
```
Now only class `Card` knows every **card selector**, so you need to encapsulate the likes logics inside its methods and call them in `index.js`. It’s a violation of **Single responsibility** principle

gennady-bars reviewed (commented):
Everything is almost fine but:


* The avatar input is empty but the button is active http://joxi.ru/E2pZKygHjWjv1r 
* The code of `ok` checking is repeated in **every** request of the project. To avoid it you need to make a special method `_checkResponse` that will check any response from the server if it’s ok or not.
* You need to close popups only in blocks `then` after a **successful response** from the server so that when there is an error it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well
* You need to return the **default** button text only in blocks `finally` so that the button could get the text back in any case: `success` or `failure`.
* So, if you call `closeAfterSubmit` in `finally`, then you need to delete `closing` and resetting, `toggleButtonState` from it (they should be in `then`)
* Now only class `Card` knows every **card selector**, so you need to encapsulate the likes logics inside its methods and call them in `index.js`. It’s a violation of **Single responsibility** principle

Aivaxela committed:
correct .then and .finally logics, refactor getUserInfo call, encapsulate card likes

Aivaxela committed:
update validity in avatar change popup

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,65 @@
+      headers: this._headers,
+      method: method,
+      body: body,
+    });
```
**Could be improved**

Please, add `.then((res) => this._checkResponse(res))` here not to duplicate it in each request


```js
  _request(endpoint, { method, body }) {
    return fetch(`${this._baseUrl}${endpoint}`, {
      headers: this._headers,
      method: method,
      body: body,
    }).then((res) => this._checkResponse(res))
  }

```


So, you will have a very neat code in `Api`


```js
  uploadCard({ Title, Link }) {
    return this._request(`cards`, {
      method: "POST",
      body: JSON.stringify({
        name: Title,
        link: Link,
      }),
    })
  }

  deleteCard(id) {
    return this._request(`cards/${id}`, {
      method: "DELETE",
    })
  }

  updateCardLike(id, method) {
    return this._request(`cards/${id}/likes`, {
      method: method,
    })
  }


```

Aivaxela replied:
🤦‍♂️

gennady-bars commented on src/pages/index.js:
```diff
@@ -2,86 +2,168 @@
+profileAvatarEdit.addEventListener("click", () => {
+  avatarEditPopup.open();
+  avatarEditFormValidator.toggleButtonState();
+  avatarEditFormValidator.resetValidation();
```
`avatarEditFormValidator.resetValidation();` is not needed when you open the avatar form because you don't need to check validity and show errors under the input on opening 


Actually, you should add `this.toggleButtonState();` inside `FormValidator` inside method `enableValidation` to initially disable the button not to duplicate it on opening 


```js
  enableValidation() {
    this._setEventListeners();
    this.toggleButtonState();
  }

```


So, you don't need anything else here in `index.js` 


```js
profileAvatarEdit.addEventListener("click", () => {
  avatarEditPopup.open();
});

```

gennady-bars reviewed (commented):
One more step and everything will be fine:


* `avatarEditFormValidator.resetValidation();` is not needed when you open the avatar form because you don't need to check validity and show errors under the empty input on opening

Aivaxela committed:
refactor some validation and api logic

gennady-bars reviewed (approved):
You’ve done a great job.

Your project has been accepted. 

Good luck
